### ISSUE TITLE
WebSocket writing prevents Idle timer timeout

### ISSUE THREAD

**Issue by Stimzz** [eclipse-vertx/vert.x#3507]
### Questions
Sending data on a WebSocket client prevents the idleTimeout from triggering. According to the Vertx dock, Setting idleTimeout on HttpClientOptions should trigger a connection closed if:

> "This determines if a connection will timeout and be closed if no data is **received** within the timeout."

https://vertx.io/docs/apidocs/io/vertx/core/http/HttpClientOptions.html

However to me it looks like the behavior is if no data is neither received nor sent. 

In the Vertx NetClientImpl class in the initChannel method on row 89:

    if (idleTimeout > 0) {
      pipeline.addLast("idle", new IdleStateHandler(0, 0, idleTimeout, idleTimeoutUnit));
    }

Looking through the Netty dock for the IdleStateHandler class:

This Netty IdleStateHandler constructor is used by NetClientImpl:

```
public IdleStateHandler(long readerIdleTime,
                        long writerIdleTime,
                        long allIdleTime,
                        java.util.concurrent.TimeUnit unit)
```

I.e. reader idle time and write idle time is set to 0 and all Idle Timeout is set to the specified value which comes from the HttpClientOption. This should then result in that either writing or reading from the channel will reset the timer and prevent timeout.

https://netty.io/4.0/api/io/netty/handler/timeout/IdleStateHandler.html

Changing the constructor call in the NetClientImpl class to:

    if (idleTimeout > 0) {
      pipeline.addLast("idle", new IdleStateHandler(idleTimeout, 0, 0, idleTimeoutUnit));
    }

Should probably solve it I assume.

### Version

3.8.4

### Context

We had a production incident where we didn't understand that the peer WebSocket server was down. This seams to have been triggered by that we send WebSocket ping frames that then prevents the idle timeout from triggering. We didn't validate that we were actually receiving WebSocket pong frames since we thought we didn't need to with the idle timeout set.

### Do you have a reproducer?

Yes, it is not 100%, we believe our production issue was caused be a dropped tcp session on the network and not that we sent data as in the unit test but according to the Netty code this test should trigger the same behavior:

https://github.com/Stimzz/vertx-idle-timeout-reproducer

Either writing data or sending ping frames (which was what we were doing) prevents the test from succeeding.

### Steps to reproduce

1. clone the repo
2. run the test by ./gradlew test
3. The test will fail after 4 seconds from a unit test timeout

### Extra

OS: Ubuntu 20
Java: 13
JVM: adoptopenjdk-13

**Comment by vietj** [comment]
WebSocket does not use NetSocket (although it could in the future), it has its own channel configuration, do you mind updating your issue ?

**Comment by Stimzz** [comment]
Sorry about that. This is the first time I am looking through the code base and I do feel lost to be perfectly honest. I am guessing and please correct me that you are referring to HttpChannelConnector?

In that case the relevant part would be the applyHttp1xConnectionOptions and applyHttp2ConnectionOptions methods of HttpChannelConnector.

After grep:ing some more I see that IdelStateHandler is instantiated in a few different places in the same way, HttpChannelConnector, HttpServerChannelInitializer, NetClientImpl and NetServerImpl (They all are configured by different option classes but said classes inherit from TCPSSLOptions). So they all use the same wording in the doc:s, i.e. the word "received" while my understanding is that the code result in any traffic (send / recv). 

I am not sure how you want to approach this, maybe just updating the doc is one solution to say any data more similar to the Netty doc for IdelStateHandler.

I remember when I wrote the initial offending code in our project that I specifically looked at and it said received as that was what I wanted. I then proceeded to never implement logic to validate that we were actually getting pong frames on our WebSockets but it turns out I should have.

What I am trying to say is that updating the new IdelStateHandler constructor calls (provided I have understood it correctly) although resulting in the code behaving as documented but maybe it will change the behavior too much in too many places? In which case it could be easier to just update the doc.

I will update the initial bug report to reflect that it is actually HttpChannelConnector that was the specific issue I had with.

**Comment by vietj** [comment]
@Stimzz would two timeout one for read and one for write solve your problem ?

**Comment by vietj** [comment]
I think we should consider having read/write idle timeout in addition of this global setting

**Comment by Stimzz** [comment]
@vietj yes and I agree. It would make the behavior more explicit from the arguments and I can see use cases where having different timeouts for read and write is desired.

**Comment by vietj** [comment]
we provide now in 4.2.0 read/write idle timeouts
